src/app/CommandSender.h

```diff
@@ -114,6 +114,62 @@ class CommandSender final : public Messaging::ExchangeDelegate
         virtual void OnDone(CommandSender * apCommandSender) = 0;
     };
 
+    // CommandSender methods are called outside of SDK. By having parameters passed using this
+    // struct we centralize defaults within the struct instead of the function's declaration
+    // parameters list.
+    struct ConfigParameters
+    {
+        ConfigParameters & SetRemoteMaxPathsPerInvoke(uint16_t aRemoteMaxPathsPerInvoke)
+        {
+            mRemoteMaxPathsPerInvoke = aRemoteMaxPathsPerInvoke;
+            return *this;
+        }
+
+        // If mRemoteMaxPathsPerInvoke is 1, this will allow the CommandSender client to contain only one command and
+        // doesn't enforce other batch commands requirements.
+        uint16_t mRemoteMaxPathsPerInvoke = 1;
+    };
+
+    // CommandSender methods are called outside of SDK. By having parameters passed using this
+    // struct we centralize defaults within the struct instead of the function's declaration
+    // parameters list.
+    struct AdditionalCommandParameters
+    {
+        // gcc bug requires us to have the constructor below
+        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96645
+        AdditionalCommandParameters() {}
+
+        AdditionalCommandParameters & SetStartOrEndDataStruct(bool aStartOrEndDataStruct)
+        {
+            mStartOrEndDataStruct = aStartOrEndDataStruct;
+            return *this;
+        }
+
+        Optional<uint16_t> mCommandRef;
+        bool mStartOrEndDataStruct = false;
+    };
+
+    // CommandSender methods are called outside of SDK. By having parameters passed using this
+    // struct we centralize defaults within the struct instead of the function's declaration
+    // parameters list.
+    struct AdditionalInvokeResponseElements
+    {
+        Optional<uint16_t> mCommandRef;
+    };
+
+    /**
+     * Called from CommandSender::Callback::OnResponse to get additional elements in InvokeResponseMessage.
+     *
+     * For legacy reasons we cannot change arguments passed through `OnResponse`, but we wanted to provide a
```

Comment:
Well, hold on.  We could add a new virtual method on Callback, have the default implementation of that call the old method, give the old method a default implementation that's a no-op so new consumers don't have to implement it.

Is there a reason that does not work?  That seems like it would be much clearer for API consumers than this totally undocumented (in the Callback parts they are supposed to look at) mechanism.

Reply:
For discussion sake, lets call this new method `NewOnResponse`. Would it be enough to specify in the description that a callback should only override one of the two callbacks, either `OnResponse` or `NewOnResponse`, but not both. The only way I figured this could be enforced is if we had `NewOnResponse` return CHIP_ERROR (instead of `void`, which is what is the return value for all these callbacks), where if we get `CHIP_ERROR_NOT_IMPLEMENTED` we call the old implementation?

Reply:
I would be fine with having something like this:
```
    // Override NewOnResponse or OnResponse but not both.
   virtual void OnResponse(args) {}
   virtual void NewOnResponse(args) { OnResponse(args); }
```
But also: you can just also call the new thing OnResponse, no?  If it has different arguments, it's not a problem (as long as we don't have anyone manually implementing the vtables COM-style).

Reply:
In trying to keep the name the same, I have [hit an issue I wasn't expecting](https://github.com/project-chip/connectedhomeip/actions/runs/7105510615/job/19342857939?pr=30675):

```
INFO    In file included from ../../examples/chip-tool/third_party/connectedhomeip/src/app/tests/suites/commands/interaction_model/InteractionModel.cpp:19:
INFO    ../../examples/chip-tool/third_party/connectedhomeip/src/app/tests/suites/commands/interaction_model/InteractionModel.h:584:18: error: 'InteractionModel::OnResponse' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
INFO      584 |     virtual void OnResponse(const chip::app::StatusIB & status, chip::TLV::TLVReader * data) = 0;
INFO          |                  ^
INFO    ../../examples/chip-tool/third_party/connectedhomeip/src/app/CommandSender.h:96:22: note: hidden overloaded virtual function 'chip::app::CommandSender::Callback::OnResponse' declared here: different number of parameters (5 vs 2)
INFO       96 |         virtual void OnResponse(CommandSender * apCommandSender, const ConcreteCommandPath & aPath, const StatusIB & aStatusIB,
INFO          |        
```

I don't know how to solve it for this case without using a different name https://github.com/project-chip/connectedhomeip/actions/runs/7105510603/job/19342859126?pr=30675
```
../../src/app/tests/integration/chip_im_initiator.cpp:196:10: error: '(anonymous namespace)::MockInteractionModelApp::OnResponse' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
  196 |     void OnResponse(const chip::app::WriteClient * apWriteClient, const chip::app::ConcreteDataAttributePath & path,
      |          ^
../../src/app/CommandSender.h:96:22: note: hidden overloaded virtual function 'chip::app::CommandSender::Callback::OnResponse' declared here: different number of parameters (5 vs 3)
   96 |         virtual void OnResponse(CommandSender * apCommandSender, const ConcreteCommandPath & aPath, const StatusIB & aStatusIB,
      |                      ^
1 error generated.
```